apps/music/js/metadata/ogg.js

```diff
@@ -39,48 +43,78 @@ var OggMetadata = (function() {
-    var header = readPageHeader(page);
-    if (header.segment_table.length !== 1) {
-      throw new Error(
-        'ogg identification header expected as only packet of first page'
-      );
-    }
-    // Skip over the identification header.
-    page.advance(header.segment_table[0]);
+  function VorbisCommentParser() {
+    this.continued = false;
+    this.read_comments = 0;
+    return this;
   }
 
-  /**
-   * Read the comment header of an Ogg container.
-   *
-   * @param {BlobView} page The audio file being parsed.
-   * @return {Promise} A Promise that resolves with the completed metadata.
-   */
-  function readCommentHeader(page) {
-    var metadata = {};
-    var header = readPageHeader(page);
-
-    var sum = function(a, b) { return a + b; };
-    var comment_length = Array.reduce(header.segment_table, sum, 0);
-
-    return new Promise((resolve, reject) => {
-      page.getMore(page.index, comment_length, (fullpage, err) => {
-        if (err) {
-          reject(err);
-          return;
-        }
+  VorbisCommentParser.prototype = {
+    /**
+     * Read the identification header of an Ogg container. It's always the first
+     * page.
+     *
+     * @param {BlobView} blobview The audio file being parsed.
+     */
+    readIdentificationHeader: function(blobview) {
+      var header = this._readPageHeader(blobview);
+      if (header.segment_table.length !== 1) {
+        throw new Error(
+          'ogg identification header expected as only packet of first page'
+        );
+      }
+      // Skip over the identification header.
+      blobview.advance(header.segment_table[0]);
+    },
+
+    /**
+     * Read the Vorbis Comment. Deal with multiple pages. Parse the
+     * picture block too.
+     *
+     * @param {BlobView} page The audio file being parsed.
+     * @return {Promise} A Promise that resolves with the completed metadata.
+     */
+    readVorbisComment: function(blobview) {
+      return new Promise((resolve, reject) => {
+
+        this._readCommentHeader(blobview).then((metadata) => {
+          LazyLoader.load('js/metadata/vorbis_picture.js').then(() => {
+            return VorbisPictureComment.parsePictureComment(metadata);
+          }).then(resolve).catch(reject);
+        });
+      });
+    },
+
+    /**
+     * Read the comment header of an Ogg container.
+     *
+     * @param {BlobView} bloview The audio file being parsed.
+     * @param {Metadata} metadata The (partially filled-in) metadata object.
+     */
+    _readCommentHeader: function(blobview) {
+      return new Promise((resolve, reject) => {
+        var metadata = {};
+        var header = this._readPageHeader(blobview);
+
+        var sum = function(a, b) { return a + b; };
+        this.page_length = Array.reduce(header.segment_table, sum, 0);
+
+        blobview.getMore(blobview.index, this.page_length, (fullpage, err) => {
+          if (err) {
+            reject(Error(err));
+          }
 
-        try {
```

Comment:
As an example, some of the `read` functions below could throw if we hit EOF, which could happen for invalid Ogg files.